### PR TITLE
Revise spack environments

### PR DESCRIPTION
**Pull request by rasolca**
Spack changed the way in which providers are selected, which impact our scalapack CI configs.

E.g. https://gitlab.com/cscs-ci/ci-testing/webhook-ci/mirrors/4700071344751697/7514005670787789/-/jobs/5602008308 where netlib-scalapack is used instead of MKL.

### CODE REVIEW

**Comment by rasolca** [comment]
cscs-ci run

File: ci/docker/debug-cpu-stdexec.yaml [hunk 1]
```diff
@@ -18,9 +18,18 @@ spack:
       true
 
   specs:
-    - dla-future@master +miniapps +ci-test +ci-check-threads ^intel-mkl threads=openmp ^mpich ^pika+stdexec ^stdexec@git.48c52df0f81c6151eecf4f39fa5eed2dc0216204=main
+    - dla-future@master build_type=Debug +miniapps +ci-test +ci-check-threads
```
Comment (RMeli) [inline ci/docker/debug-cpu-stdexec.yaml:21]: In `common.yaml`, `threads=openmp` is only a variant of `intel-mkl`. My current understanding is that variants have low precedence. Do we need
```
intel-mkl:
  require:
  - "threads=openmp"
```
to ensure the same behavior?

File: ci/docker/debug-cpu.yaml [hunk 2]
```diff
@@ -18,15 +18,22 @@ spack:
...
+    - dla-future@master build_type=Debug +miniapps +ci-test +ci-check-threads
 
   packages:
     all:
       variants:
         - 'build_type=Release'
+    blas:
+      require:: 'openblas'
+    lapack:
+      require:: 'openblas'
```
Comment (RMeli) [inline ci/docker/debug-cpu.yaml:30]: Related to comment above for `intel-mkl`.
```suggestion
    blas:
      require: 'openblas threads=openmp'
    lapack:
      require: 'openblas threads=openmp'
```
?

File: ci/docker/debug-cuda-scalapack.yaml [hunk 3]
```diff
@@ -18,13 +18,22 @@ spack:
...
+    - dla-future@master build_type=Debug +cuda +miniapps +scalapack +ci-test +hdf5
 
   packages:
     all:
       variants:
         - 'build_type=Release'
+    blas:
+      require:: 'openblas'
+    lapack:
+      require:: 'openblas'
```
Comment (RMeli) [inline ci/docker/debug-cuda-scalapack.yaml:30]:
```suggestion
    blas:
      require: 'openblas threads=openmp'
    lapack:
      require: 'openblas threads=openmp'
```
?

File: ci/docker/debug-cuda.yaml [hunk 4]
```diff
@@ -18,15 +18,17 @@ spack:
       true
 
   specs:
-    - dla-future@master build_type=Debug +cuda +miniapps +ci-test ^openblas threads=openmp ^mpich
+    - dla-future@master build_type=Debug +cuda +miniapps +ci-test
```
Comment (RMeli) [inline ci/docker/debug-cuda.yaml:21]: Is
```
    blas:
      require: 'openblas threads=openmp'
    lapack:
      require: 'openblas threads=openmp'
```
missing in this YAML file? I think this now defaults to `intel-mkl`.

Comment (rasolca) [reply ci/docker/debug-cuda.yaml:21]: Ups...

**Comment by rasolca** [comment]
@RMeli Let me answer all your comments here:
I had a chat with @albestro discussing the best method and we concluded the following:

1) the virtual packages just specify which provider you want to use. The options for the specific packages go in the specs of the package itself. In this way we can specify `threads=openmp` for both openblas and intel-mkl in `common.yaml`

2) use array instead of string in `require:` (only exception in virtual packages)
in this way specs combine and are not overwritten.

3) prefer `variant:` in `common.yaml` unless it is an hard requirement (e.g. mpich version 3.4.2 is an hard requirement otherwise sarus mpi replacement is not working.)
In the CI, `variant:` should not be problematic as we always start with a fresh spack.
Moreover, being just a preference it is possible for the environments to specify a requirement which is in contrast with the `common.yaml` defaults.

**Comment by RMeli** [comment]
> The options for the specific packages go in the specs of the package itself.

Ok, I see what it is going on now, thanks for the clarification.

> In the CI, `variant:` should not be problematic as we always start with a fresh spack.

Sounds good.

Thanks for the clarifications!

File: ci/docker/debug-cuda.yaml (hunk 4 above)
Comment (RMeli) [reply ci/docker/debug-cuda.yaml:21]: This is now fixed.

**Comment by rasolca** [comment]
cscs-ci run

**Review by msimberg — APPROVED** [review]

**Review by albestro — APPROVED** [review]

**Comment by rasolca** [comment]
cscs-ci run

**Review by albestro — APPROVED** [review]

**Comment by albestro** [comment]
Good catch @rasolca 🧐 😄

**Pull request merged**